streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -17,242 +17,111 @@
+        this.streamsKafkaClient = streamsKafkaClient;
         this.replicationFactor = replicationFactor;
         this.windowChangeLogAdditionalRetention = windowChangeLogAdditionalRetention;
     }
 
-    public void makeReady(InternalTopicConfig topic, int numPartitions) {
-        boolean topicNotReady = true;
-
-        while (topicNotReady) {
-            Map<Integer, List<Integer>> topicMetadata = getTopicMetadata(topic.name());
+    public void makeReady(final InternalTopicConfig topic, int numPartitions) {
+        Map<InternalTopicConfig, Integer> topics = new HashMap<>();
+        topics.put(topic, numPartitions);
+        makeReady(topics);
+    }
 
-            if (topicMetadata == null) {
-                try {
-                    createTopic(topic, numPartitions, replicationFactor);
-                } catch (ZkNodeExistsException e) {
-                    // ignore and continue
-                }
-            } else {
-                if (topicMetadata.size() > numPartitions) {
-                    // else if topic exists with more #.partitions than needed, delete in order to re-create it
-                    try {
-                        deleteTopic(topic.name());
-                    } catch (ZkNodeExistsException e) {
-                        // ignore and continue
-                    }
-                } else if (topicMetadata.size() < numPartitions) {
-                    // else if topic exists with less #.partitions than needed, add partitions
-                    try {
-                        addPartitions(topic.name(), numPartitions - topicMetadata.size(), replicationFactor, topicMetadata);
-                    } catch (ZkNoNodeException e) {
-                        // ignore and continue
-                    }
-                } else {
-                    topicNotReady = false;
-                }
+    /**
+     * Prepares the set of given internal topics. If the topic with the correct number of partitions exists ignores it. For the ones with different number of
+     * partitions delete them and create new ones with correct number of partitons along with the non existing topics.
+     * @param topics
+     */
+    public void makeReady(final Map<InternalTopicConfig, Integer> topics) {
+
+        Collection<MetadataResponse.TopicMetadata> topicMetadatas = streamsKafkaClient.fetchTopicMetadata();
+        Map<InternalTopicConfig, Integer> topicsToBeDeleted = getTopicsToBeDeleted(topics, topicMetadatas);
+        Map<InternalTopicConfig, Integer> topicsToBeCreated = filterExistingTopics(topics, topicMetadatas);
+        topicsToBeCreated.putAll(topicsToBeDeleted);
+        streamsKafkaClient.deleteTopics(topicsToBeDeleted);
+        for (int i = 0; i < MAX_TOPIC_READY_TRY; i++) {
+            try {
+                streamsKafkaClient.createTopics(topicsToBeCreated, replicationFactor, windowChangeLogAdditionalRetention);
+                return;
+            } catch (StreamsException ex) {
```

Comment:
Better add a log entry here since otherwise the error message in `StreamsException` will never be shown anywhere.